Sources/NIO/BaseSocketChannel.swift

```diff
@@ -14,6 +14,190 @@
 
 import NIOConcurrencyHelpers
 
+struct SocketChannelLifecycleManager {
+    // MARK: Types
+    enum State {
+        case neverRegistered
+        case registeredNeverActivated
+        case registered
+        case activated
+        case closed
+    }
+
+    enum Event {
+        case activate
+        case deactivate
+        case register
+        case close
+    }
+
+    // MARK: properties
+    // this is queried from the Channel, ie. must be thread-safe
+    internal let isActiveAtomic: Atomic<Bool>
+    // these are only to be accessed on the EventLoop
+    internal let channelPipeline: ChannelPipeline
+    private let eventLoop: EventLoop
+    private var currentState: State = .neverRegistered {
+        didSet {
+            assert(self.eventLoop.inEventLoop)
+            switch (oldValue, self.currentState) {
+            case (_, .activated):
+                self.isActiveAtomic.store(true)
+            case (.activated, _):
+                self.isActiveAtomic.store(false)
+            default:
+                ()
+            }
```

Comment:
I assume the reason this conditional is a bit complex is to avoid doing too many atomic ops?

Reply:
@Lukasa indeed, also don't think it's too complex tbh